proposals/nullable-reference-types-specification.md

```diff
@@ -0,0 +1,266 @@
+﻿# Nullable Reference Types Specification
+
+*** This is a work in progress - several parts are missing or incomplete. ***
+
+# Syntax
+
+## Nullable reference types
+
+Nullable reference types have the same syntax `T?` as the short form of nullable value types, but do not have a corresponding long form.
+
+For the purposes of the specification, the current `nullable_type` production is renamed to `nullable_value_type`, and a `nullable_reference_type` production is added:
+
+```antlr
+reference_type
+    : ...
+    | nullable_reference_type
+    ;
+    
+nullable_reference_type
+    : non_nullable_reference_type '?'
+    ;
+    
+non_nullable_reference_type
+    : type
+    ;
+```
+
+The `non_nullable_reference_type` in a `nullable_reference_type` must be a non-nullable reference type (class, interface, delegate or array), or a type parameter that is constrained to be a non-nullable reference type (through the `class` constraint, or a class other than `object`).
+
+Nullable reference types cannot occur in the following positions:
+
+- as a base class or interface
+- as the receiver of a `member_access`
+- as the `type` in an `object_creation_expression`
+- as the `delegate_type` in a `delegate_creation_expression`
+- as the `type` in an `is_expression`, a `catch_clause` or a `type_pattern`
+- as the `interface` in a fully qualified interface member name
+
+A warning is given on a `nullable_reference_type` where the nullable annotation context is disabled.
+
+## Nullable class constraint
+
+The `class` constraint has a nullable counterpart `class?`:
+
+```antlr
+primary_constraint
+    : ...
+    | 'class' '?'
+    ;
+```
+
+## Null-forgiving expressions
+
+The post-fix `!` operator is called the null-forgiving operator.
```

Comment:
'Null-forgiving' hints at smoothing over the effect of nulls. It sounds like an odd term for something that will blow up at runtime if there is a null.